[Form] Added support for caching choice lists based on options

HeahDude opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | ~
| License       | MIT
| Doc PR        | symfony/symfony-docs#13182

Currently, the `CachingFactoryDecorator` is responsible for unnecessary memory usage, anytime a choice option is set with a callback option defined as an anonymous function or a loader, then a new hash is generated for the choice list, while we may expect the list to be reused once "finally" configured in a form type or choice type extension.

A simple case is when using one of the core intl choice types in a collection:
```php
// ...
class SomeFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('some_choices', ChoiceType::class, [
                // before: cached choice list (unnecessary overhead)
                // after: no cache (better perf)
                'choices' => $someObjects,
                'choice_value' => function (?object $choice) { /* return some string */ },
            ])

            // see below the nested effects
            ->add('nested_fields', CollectionType::class, [
                'entry_type' => NestedFormType::class,
            ])
    // ...
}

// ...
class NestedFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            // ...
            ->add('some_other_choices', ChoiceType::class, [
                // before: cached choice list for every entry because we define a new closure instance for each field
                // after: no cache, a bit better for the same result, but much better if it were not nested in a collection
                'choices' => $someOtherObjects,
                'choice_value' => function (?object $otherChoice) { /* return some string */ },
            ])

            ->add('some_loaded_choices', ChoiceType::class, [
                // before: cached but for every entry since every field will have its
                //         own instance of loader, generating a new hash 
                // after: no cache, same pro/cons as above
                'choice_loader' => new CallbackChoiceLoader(function() { /* return some choices */}),
                // or 
                'choice_loader' => new SomeLoader(),
            ])

            ->add('person', EntityType::class, [
                // before: cached but for every entry, because we define extra `choice_*` option
                // after: no cache, same pro/cons as above
                'class' => SomeEntity::class,
                'choice_label' => function (?SomeEntity $choice) { /* return some label */},
            ])

            // before: cached for every entry, because the type define some "choice_*" option
            // after: cached only once, better perf since the same loader is used for every entry
            ->add('country', CountryType::class)

            // before: cached for every entry, because the type define some "choice_*" option
            // after: no cache, same pro/cons as above
            ->add('locale', LocaleType::class, [
                'preferred_choices' => [ /* some preferred locales */ ],
                'group_by' => function (?string $locale, $label) { /* return some group */ },
            ])
// ...
```

In such cases, we would expect every entries to use the same cached intl choice list, but not, as many list and views as entries will be kept in cache. This is even worse if some callback options like `choice_label`, `choice_value`, `choice_attr`, `choice_name`, `preferred_choices` or `group_by` are used.
This PR helps making cache explicit when needed and ~deprecate~ drop unexpected implicit caching of choice list for most simple cases responsible of unnecessary overhead.

The result is better performance just by upgrading to 5.1 \o/.
But to solve the cases above when cache is needed per options, one should now use the new `ChoiceList` static methods to wrap option values, which is already done internally in this PR.

```php
use Symfony\Component\Form\ChoiceList\ChoiceList;
// ...
class NestedFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            // explicitly shared cached choice lists between entries

            ->add('some_other_choices', ChoiceType::class, [
                'choices' => $someOtherObjects,
                'choice_value' => ChoiceList::value($this, function (?object $otherChoice) {
                    /* return some string */ 
                }),
            ]),

            ->add('some_loaded_choices', ChoiceType::class, [
                'choice_loader' => ChoiceList::lazy($this, function() {
                    /* return some choices */
                })),
                // or
                'choice_loader' => ChoiceList::loader($this, new SomeLoader()),
            ]),

            ->add('person', EntityType::class, [
                'class' => SomeEntity::class,
                'choice_label' => ChoiceList::label($this, function (?SomeEntity $choice) {
                    /* return some label */
                },
            ])

            // nothing to do :)
            ->add('country', CountryType::class)

            ->add('locale', LocaleType::class, [
                'preferred_choices' => ChoiceList::preferred($this, [ /* some preferred locales */ ]),
                'group_by' => ChoiceList::groupBy($this, function (?string $locale, $label) {
                    /* return some group */
                }),
            ])
// ...
```

I've done some nice profiling with Blackfire and the simple example above in a fresh website skeleton and only two empty entries as initial data, then submitting an empty form. That gives the following results:

 * Rendering the form - Before vs After

  <img width="714" alt="Screenshot 2020-02-16 at 9 24 58 PM" src="https://user-images.githubusercontent.com/10107633/74612132-de533180-5102-11ea-9cc4-296a16949d90.png">

 * Rendering the form - Before vs After with `ChoiceList` helpers

  <img width="670" alt="Screenshot 2020-02-16 at 9 26 51 PM" src="https://user-images.githubusercontent.com/10107633/74612155-122e5700-5103-11ea-9c16-5d80a7541f4b.png">

 * Submitting the form - Before vs After

  <img width="670" alt="Screenshot 2020-02-16 at 9 28 01 PM" src="https://user-images.githubusercontent.com/10107633/74612172-3be77e00-5103-11ea-9a18-4294e05402d2.png">

 * Submitting the form - Before vs After with `ChoiceList` helpers

  <img width="670" alt="Screenshot 2020-02-16 at 9 29 10 PM" src="https://user-images.githubusercontent.com/10107633/74612193-689b9580-5103-11ea-86b9-5b4906200021.png">

_________

TODO:
- [x] Docs
- [x] More profiling
- [x] Add some tests

#EUFOSSA

HeahDude committed:
[Form] Added support for caching choice lists based on options

OskarStark commented on src/Symfony/Component/Form/ChoiceList/ChoiceList.php:
```diff
@@ -0,0 +1,101 @@
+{
+    /**
+     * @param FormTypeInterface     $formType A form type configuring a list
+     * @param ChoiceLoaderInterface $loader   Any pseudo callable to create a unique value from a choice
```
Wrong php doc

OskarStark commented on src/Symfony/Component/Form/ChoiceList/ChoiceList.php:
```diff
@@ -0,0 +1,101 @@
+     * @param callable          $label    Any pseudo callable to create a label from a choice
+     * @param mixed             $vary     Dynamic data used to compute a unique hash when caching the callback
+     */
+    public static function label(FormTypeInterface $formType, $label, $vary = null): ChoiceLabel
```
If $label is a callable (phpdoc) can you add a type hint here like above?

OskarStark replied:
Same for the following in the other methods

HeahDude replied:
No we can't, a string may be decorated to be a property path, hence "pseudo" callable description

OskarStark replied:
Thanks for the explanation this makes sense

OskarStark commented on src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php:
```diff
@@ -87,10 +89,18 @@ public function createListFromChoices($choices, $value = null)
+            $value = $value->getCallback();
+            $hash = self::generateHash([$choices, $value], 'fromChoicesValueCallback');
+        } elseif ($value) {
+            @trigger_error(sprintf('Not passing an instance of "%s" to create choice list cache per callback as been deprecated since 4.3. Only choices will be considered in 5.0.', Cache\ChoiceValue::class), E_USER_DEPRECATED);
```
as -> has

OskarStark commented on src/Symfony/Component/Form/Extension/Core/Type/ChoiceType.php:
```diff
@@ -324,13 +331,13 @@ public function configureOptions(OptionsResolver $resolver)
-        $resolver->setAllowedTypes('choice_attr', ['null', 'array', 'callable', 'string', 'Symfony\Component\PropertyAccess\PropertyPath']);
-        $resolver->setAllowedTypes('preferred_choices', ['array', '\Traversable', 'callable', 'string', 'Symfony\Component\PropertyAccess\PropertyPath']);
-        $resolver->setAllowedTypes('group_by', ['null', 'callable', 'string', 'Symfony\Component\PropertyAccess\PropertyPath']);
+        $resolver->setAllowedTypes('choice_loader', ['null', 'Symfony\Component\Form\ChoiceList\Loader\ChoiceLoaderInterface', ChoiceLoader::class]);
```
You could use ::class constant for the others, too

HeahDude replied:
Unwanted merge conflicts with 3.4 I guess :)

vudaltsov commented on src/Symfony/Bridge/Doctrine/CHANGELOG.md:
```diff
@@ -7,6 +7,7 @@ CHANGELOG
  * changed guessing of DECIMAL to set the `input` option of `NumberType` to string
  * deprecated not passing an `IdReader` to the `DoctrineChoiceLoader` when query can be optimized with a single id field
  * deprecated passing an `IdReader` to the `DoctrineChoiceLoader` when entities have a composite id
+ * added a `ChoiceList` facade to explicit choice lists caching based on callback options or loader when configured by specific form types
```
```suggestion
 * added a `ChoiceList` facade for explicit choice lists caching based on callback options or loader when configured by specific form types
```

HeahDude replied:
This is not really the same meaning:
"I added something **to explicit** something when being something"
"I added something for **explicit something** when being something"

I would prefer to keep the first.

nicolas-grekas replied:
I would ask a native english speaker here :)

vudaltsov replied:
`explicit` is an adjective, not a verb. So you cannot `to explicit smth`
https://dictionary.cambridge.org/dictionary/english-russian/explicit

HeahDude replied:
Indeed. It is in French and wrongly translated as a verb here http://www.wordreference.com/fren/expliciter. I would propose `added a `ChoiceList` facade to make explicit choice lists caching...` then.

vudaltsov replied:
```suggestion
 * added a `ChoiceList` facade to leverage explicit choice list caching based on callback options or specific loaders
```

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/ChoiceList.php:
```diff
@@ -0,0 +1,101 @@
+     * @param ChoiceLoaderInterface $loader   Any pseudo callable to create a unique value from a choice
+     * @param mixed                 $vary     Dynamic data used to compute a unique hash when caching the loader
+     */
+    public static function lazy(FormTypeInterface $formType, callable $choices, $vary = null): ChoiceLoader
```
Let's add a description for each method. Since this class is supposed to be used in user land, it should be really comprehensible.

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/AbstractStaticCallback.php:
```diff
@@ -0,0 +1,59 @@
+     * @param callable          $callback Any pseudo callable to create choice list property
+     * @param mixed             $vary     Dynamic data used to compute a unique hash when caching the callback
+     */
+    final public function __construct(FormTypeInterface $formType, $callback, $vary = null)
```
Why not
```suggestion
    final public function __construct(FormTypeInterface $formType, callable $callback, $vary = null)
```
?

vudaltsov replied:
Saw the comment above to late, sorry.

vudaltsov replied:
Let's then be explicit about it? `callable|string Any callable or property path to create choice list property`

HeahDude replied:
Actually we don't know what decorator will be used here, each decorator provides its own consistent phpdocs, but those properties are never type hinted to be used in factories methods.
Furthermore, userland and third party code can also decorate the factory.

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/AbstractStaticCallback.php:
```diff
@@ -0,0 +1,59 @@
+        return $this->callback;
+    }
+
+    public static function reset()
```
```suggestion
    public static function reset(): void
```

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/AbstractStaticCallback.php:
```diff
@@ -0,0 +1,59 @@
+    /**
+     * @return mixed
+     */
+    final public function getCallback()
```
```suggestion
    final public function getCallback(): callable
```

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/ChoiceLoader.php:
```diff
@@ -0,0 +1,82 @@
+     * @param ChoiceLoaderInterface $loader   A static loader used by the type
+     * @param mixed                 $vary     Dynamic data used to compute a unique hash when caching the loader
+     */
+    final public function __construct(FormTypeInterface $formType, ChoiceLoaderInterface $loader, $vary)
```
Why `$vary` is optional in `AbstractStaticCallback` and required here?
```suggestion
    final public function __construct(FormTypeInterface $formType, ChoiceLoaderInterface $loader, $vary = null)
```

HeahDude replied:
A mistake, thanks!

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/ChoiceLoader.php:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * Not implemented.
```
`Not implemented.` — what does it mean?

HeahDude replied:
In my first draft I threw bad method call exceptions. This is just a flyweight, impelling the interface is being able to decorate the loader and pass the factory method because it **is** type hinted on the contrary to callables.
Not sure whether it should really implement those though. What do you and other think?

HeahDude replied:
It is safer to implement it in a way that we cannot know for sure that the caching factory decorator will be used, if a custom one is injected instead of decorating the existing. I may remove the comments and keep it as is I guess.

HeahDude replied:
I others agree, then I'll also reimplement `__invoke` in the abstract callback flyweight, for consistency, which I removed before decoupling this commit.

HeahDude replied:
Comments removed.

vudaltsov commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/ChoiceLoader.php:
```diff
@@ -0,0 +1,82 @@
+        return $this->loader->loadValuesForChoices($value);
+    }
+
+    public static function reset()
```
```suggestion
    public static function reset(): void
```

fabpot commented:
@HeahDude What's the status of this PR? Are you willing to finish it?

HeahDude commented:
This one is ready for another round of reviews.

HeahDude commented:
I've fixed the tests, it seems that they will benefit from the optimization too in 6.0.

ping @xabbuh @stof @nicolas-grekas

nicolas-grekas commented on src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php:
```diff
@@ -101,9 +114,31 @@ public function createListFromChoices(iterable $choices, $value = null)
      * {@inheritdoc}
      */
-    public function createListFromLoader(ChoiceLoaderInterface $loader, $value = null)
+    public function createListFromLoader($loader, $value = null)
```
BC break here isn't it?

HeahDude replied:
I guess you're right... I need the handle the case where this decorator is extended 🤔,  what about annotating factories as final? They should be decorated instead.

HeahDude replied:
Fixed.

HeahDude commented:
Rebased and squashed here, the last implementation does not trigger deprecations anymore, it is just about optimizing when we can, without the cost of a BC layer.

nicolas-grekas commented on src/Symfony/Component/Form/ChoiceList/ChoiceList.php:
```diff
@@ -0,0 +1,135 @@
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+final class ChoiceList
```
I'm not sure I see the point of this class. I'd suggest removing it and inlining the instantiations instead.

HeahDude replied:
It's the user land part of this feature, everything else is about optimising the internals, many advantages to me:

- easier to document/learn, only one class instead of many (I also add the `ChoiceFilter` in #35733 and we may add more in the future like a `ChoiceLabelAttr` I'got locally)
- better discovery since every factory method has its own specific phpdoc instead of the generic constructor in the abstract class
- better auto complete, there are a lot of classes using the `Choice` word already:
  <img width="633" alt="Screenshot 2020-02-17 at 7 08 15 PM" src="https://user-images.githubusercontent.com/10107633/74677559-ef5e7a00-51b8-11ea-8cad-ca709dd864bd.png">

  VS 

  <img width="699" alt="Screenshot 2020-02-17 at 7 06 07 PM" src="https://user-images.githubusercontent.com/10107633/74677540-e66da880-51b8-11ea-85f0-9e1fd444fcc9.png">

I agree we could inline in our own usage in this PR to save a few static calls, but I wouldn't either since the types are some kind of documentation and given the Blackfire profiles I got with the current implementation.

HeahDude replied:
You can see this class as a `ChoiceList` builder or configurator

nicolas-grekas commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/AbstractStaticOption.php:
```diff
@@ -0,0 +1,64 @@
+     * @param mixed                                        $option   Any pseudo callable, array, string or bool to define a choice list option
+     * @param mixed|null                                   $vary     Dynamic data used to compute a unique hash when caching the option
+     */
+    final public function __construct($formType, $option, $vary = null)
```
why final?

HeahDude replied:
Why not? This class is responsible for clearing the cache of all options, they should not override/extends it. I see no use case currently, we have everything we need here with those  three args.

HeahDude replied:
This is a design needed by the `CachingFactoryDecorator` no matter the option.

nicolas-grekas commented on src/Symfony/Component/Form/ChoiceList/Factory/Cache/AbstractStaticOption.php:
```diff
@@ -0,0 +1,64 @@
+    /**
+     * @return mixed
+     */
+    final public function getOption()
```
I'd remove the final keywords - `@internal` is enough to express our intend, no need to be extreme

HeahDude replied:
The property is private and there is no need for extension.

nicolas-grekas commented on src/Symfony/Component/Form/ChoiceList/Factory/CachingFactoryDecorator.php:
```diff
@@ -139,5 +197,6 @@ public function reset()
     {
         $this->lists = [];
         $this->views = [];
+        Cache\AbstractStaticOption::reset();
```
why? if this is static, it shouldn't need to be reset between requests.

HeahDude replied:
It's consistent with the lists and the views, the options used to build them may be defined per request

nicolas-grekas reviewed (approved):
Ok, thanks for your answers :)

HeahDude commented:
Thank you for the review <3. I've updated the description to better explain the common scenarii and the after/before effects.
I've also done more profiling and added some nice screenshots :).
Docs PR is coming!

HeahDude commented:
Doc PR opened symfony/symfony-docs#13182.

fabpot commented:
Thank you @HeahDude.
